src/Models/BaseElement.php

```diff
@@ -1033,4 +1033,38 @@ public function EvenOdd()
 
         return  ($odd) ? 'odd' : 'even';
     }
+
+    /**
+     * Returns true if the current block state can be cached
+     *
+     * This describes the current block instance state cacheability.
+     *
+     * E.g. if the block is currently rendering a half-filled form with
+     * private user data, it's not cacheable. If the block state is going to
+     * render a simple HTML content, perhaps it may be cached.
+     *
+     * @return bool False by default
+     */
+    public function isCacheable()
+    {
+        return false;
+    }
+
+    /**
+     * Returns the block cache key
+     *
+     * The key should be unique for every instance of the block (per BaseElement->ID).
+     *
+     * It does not have to be unique across different blocks
+     * nor even different instances of the same block.
+     *
+     * If the block is not cacheable {@see self::isCacheable}, this method may
+     * return anything (e.g. non-unique values).
+     *
+     * @return string
+     */
+    public function getCacheKey()
+    {
+        return '';
```

Comment:
This default is quite dangerous. This should either provide a good default key or throw.

Reply:
Yes, I considered making it random by default and this could be dangerous if we needed it to be unique.
On the other hand, I decided it's better in this way because:
 - it's only bad if `isCacheable` returns `true`
 - if we return random here and users implement `isCacheable` but forget to override this method (dev mistake)
   - it's harder to notice the mistake in `DEV` stage
   - when it hits production, the app cache storage will be overwhelmed randomly, which is really hard to debug and catch

Reply:
Throwing an exception might be a good alternative though, I didn't think about that one.
That's a good one. I think we should do that.

Reply:
I'm just thinking about how would I use this is another project where we implemented a very similar solution. In that project most blocks work with a very simple unique key, only few need to override, vast majority is cachable, but some are not. Given your feedback my recommendation is:

* provide a good default for `getCacheKey` for example getUniqueKey() which I mentioned above
* add an extension point inside the `getCacheKey` to override this in one place, overriding this on all blocks in a project is not the best option
* provide static configuration for `isCacheable` which defaults to `false` instead of hard coding the value in the code